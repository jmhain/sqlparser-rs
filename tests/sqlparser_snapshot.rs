// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#![warn(clippy::all)]

use sqlparser::ast::Statement;
use sqlparser::dialect::{Dialect, GenericDialect};
use sqlparser::parser::Parser;
use std::path::Path;

fn test_dialect(path: &Path, dialect: &dyn Dialect) -> datatest_stable::Result<()> {
    let sql = std::fs::read_to_string(path)?;

    let stmts = Parser::parse_sql(dialect, &sql)?;
    let name = path
        .strip_prefix(Path::new("tests/queries"))?
        .to_str()
        .unwrap();
    insta::assert_yaml_snapshot!(name, stmts);

    // Rendering the AST to a string and re-parsing it should yield the same AST.
    for stmt in stmts {
        let [round_trip_stmt]: [Statement; 1] = Parser::parse_sql(dialect, &stmt.to_string())?
            .try_into()
            .map_err(|_| "expected a single statement")?;
        pretty_assertions::assert_eq!(stmt, round_trip_stmt);
    }

    Ok(())
}

fn test_generic(path: &Path) -> datatest_stable::Result<()> {
    test_dialect(path, &GenericDialect)
}

datatest_stable::harness!(test_generic, "tests/queries/generic", r"^.*.sql",);
