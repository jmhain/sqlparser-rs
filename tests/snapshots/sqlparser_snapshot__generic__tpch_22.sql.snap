---
source: tests/sqlparser_snapshot.rs
expression: stmts
---
- Query:
    with: ~
    body:
      Select:
        distinct: ~
        top: ~
        projection:
          - UnnamedExpr:
              Identifier:
                value: cntrycode
                quote_style: ~
          - ExprWithAlias:
              expr:
                Function:
                  name:
                    - value: count
                      quote_style: ~
                  args:
                    - Unnamed: Wildcard
                  filter: ~
                  null_treatment: ~
                  over: ~
                  distinct: false
                  special: false
                  order_by: []
              alias:
                value: numcust
                quote_style: ~
          - ExprWithAlias:
              expr:
                Function:
                  name:
                    - value: sum
                      quote_style: ~
                  args:
                    - Unnamed:
                        Expr:
                          Identifier:
                            value: c_acctbal
                            quote_style: ~
                  filter: ~
                  null_treatment: ~
                  over: ~
                  distinct: false
                  special: false
                  order_by: []
              alias:
                value: totacctbal
                quote_style: ~
        into: ~
        from:
          - relation:
              Derived:
                lateral: false
                subquery:
                  with: ~
                  body:
                    Select:
                      distinct: ~
                      top: ~
                      projection:
                        - ExprWithAlias:
                            expr:
                              Substring:
                                expr:
                                  Identifier:
                                    value: c_phone
                                    quote_style: ~
                                substring_from:
                                  Value:
                                    Number:
                                      - "1"
                                      - false
                                substring_for:
                                  Value:
                                    Number:
                                      - "2"
                                      - false
                                special: false
                            alias:
                              value: cntrycode
                              quote_style: ~
                        - UnnamedExpr:
                            Identifier:
                              value: c_acctbal
                              quote_style: ~
                      into: ~
                      from:
                        - relation:
                            Table:
                              name:
                                - value: customer
                                  quote_style: ~
                              alias: ~
                              args: ~
                              with_hints: []
                              version: ~
                              partitions: []
                          joins: []
                      lateral_views: []
                      selection:
                        BinaryOp:
                          left:
                            BinaryOp:
                              left:
                                InList:
                                  expr:
                                    Substring:
                                      expr:
                                        Identifier:
                                          value: c_phone
                                          quote_style: ~
                                      substring_from:
                                        Value:
                                          Number:
                                            - "1"
                                            - false
                                      substring_for:
                                        Value:
                                          Number:
                                            - "2"
                                            - false
                                      special: false
                                  list:
                                    - Value:
                                        SingleQuotedString: "13"
                                    - Value:
                                        SingleQuotedString: "31"
                                    - Value:
                                        SingleQuotedString: "23"
                                    - Value:
                                        SingleQuotedString: "29"
                                    - Value:
                                        SingleQuotedString: "30"
                                    - Value:
                                        SingleQuotedString: "18"
                                    - Value:
                                        SingleQuotedString: "17"
                                  negated: false
                              op: And
                              right:
                                BinaryOp:
                                  left:
                                    Identifier:
                                      value: c_acctbal
                                      quote_style: ~
                                  op: Gt
                                  right:
                                    Subquery:
                                      with: ~
                                      body:
                                        Select:
                                          distinct: ~
                                          top: ~
                                          projection:
                                            - UnnamedExpr:
                                                Function:
                                                  name:
                                                    - value: avg
                                                      quote_style: ~
                                                  args:
                                                    - Unnamed:
                                                        Expr:
                                                          Identifier:
                                                            value: c_acctbal
                                                            quote_style: ~
                                                  filter: ~
                                                  null_treatment: ~
                                                  over: ~
                                                  distinct: false
                                                  special: false
                                                  order_by: []
                                          into: ~
                                          from:
                                            - relation:
                                                Table:
                                                  name:
                                                    - value: customer
                                                      quote_style: ~
                                                  alias: ~
                                                  args: ~
                                                  with_hints: []
                                                  version: ~
                                                  partitions: []
                                              joins: []
                                          lateral_views: []
                                          selection:
                                            BinaryOp:
                                              left:
                                                BinaryOp:
                                                  left:
                                                    Identifier:
                                                      value: c_acctbal
                                                      quote_style: ~
                                                  op: Gt
                                                  right:
                                                    Value:
                                                      Number:
                                                        - "0.00"
                                                        - false
                                              op: And
                                              right:
                                                InList:
                                                  expr:
                                                    Substring:
                                                      expr:
                                                        Identifier:
                                                          value: c_phone
                                                          quote_style: ~
                                                      substring_from:
                                                        Value:
                                                          Number:
                                                            - "1"
                                                            - false
                                                      substring_for:
                                                        Value:
                                                          Number:
                                                            - "2"
                                                            - false
                                                      special: false
                                                  list:
                                                    - Value:
                                                        SingleQuotedString: "13"
                                                    - Value:
                                                        SingleQuotedString: "31"
                                                    - Value:
                                                        SingleQuotedString: "23"
                                                    - Value:
                                                        SingleQuotedString: "29"
                                                    - Value:
                                                        SingleQuotedString: "30"
                                                    - Value:
                                                        SingleQuotedString: "18"
                                                    - Value:
                                                        SingleQuotedString: "17"
                                                  negated: false
                                          group_by:
                                            Expressions: []
                                          cluster_by: []
                                          distribute_by: []
                                          sort_by: []
                                          having: ~
                                          named_window: []
                                          qualify: ~
                                      order_by: []
                                      limit: ~
                                      limit_by: []
                                      offset: ~
                                      fetch: ~
                                      locks: []
                          op: And
                          right:
                            Exists:
                              subquery:
                                with: ~
                                body:
                                  Select:
                                    distinct: ~
                                    top: ~
                                    projection:
                                      - Wildcard:
                                          opt_exclude: ~
                                          opt_except: ~
                                          opt_rename: ~
                                          opt_replace: ~
                                    into: ~
                                    from:
                                      - relation:
                                          Table:
                                            name:
                                              - value: orders
                                                quote_style: ~
                                            alias: ~
                                            args: ~
                                            with_hints: []
                                            version: ~
                                            partitions: []
                                        joins: []
                                    lateral_views: []
                                    selection:
                                      BinaryOp:
                                        left:
                                          Identifier:
                                            value: o_custkey
                                            quote_style: ~
                                        op: Eq
                                        right:
                                          Identifier:
                                            value: c_custkey
                                            quote_style: ~
                                    group_by:
                                      Expressions: []
                                    cluster_by: []
                                    distribute_by: []
                                    sort_by: []
                                    having: ~
                                    named_window: []
                                    qualify: ~
                                order_by: []
                                limit: ~
                                limit_by: []
                                offset: ~
                                fetch: ~
                                locks: []
                              negated: true
                      group_by:
                        Expressions: []
                      cluster_by: []
                      distribute_by: []
                      sort_by: []
                      having: ~
                      named_window: []
                      qualify: ~
                  order_by: []
                  limit: ~
                  limit_by: []
                  offset: ~
                  fetch: ~
                  locks: []
                alias:
                  name:
                    value: custsale
                    quote_style: ~
                  columns: []
            joins: []
        lateral_views: []
        selection: ~
        group_by:
          Expressions:
            - Identifier:
                value: cntrycode
                quote_style: ~
        cluster_by: []
        distribute_by: []
        sort_by: []
        having: ~
        named_window: []
        qualify: ~
    order_by:
      - expr:
          Identifier:
            value: cntrycode
            quote_style: ~
        asc: ~
        nulls_first: ~
    limit: ~
    limit_by: []
    offset: ~
    fetch: ~
    locks: []

